RTPS: msg generation: generalize scripts

TSC21 opened the pull request:
This PR generalizes the scripts fro generating the RTPS related headers and source files.
1. Allows passing the templates as arguments, so to allow the scripts to use different templates;
2. Allows passing the pkg name, which is useful when setting field types which depends on other msg types that belong to a certain package - it also allows using the scripts out of the PX4 Firmware repo;
3. Sets the RTPS msg ID's to the templates by loading these from a JSON file - this allows setting other JSON files with other IDs or message naming (which for the ROS side is required, since the msg naming is different).

LorenzMeier commented:
Great! What do we need to wrap this up so we can start to use it on a daily basis?

TSC21 commented:
Still cleaning it up, but when you talk about use it on a daily basis you mean the RTPS<->ROS bridge? If that's so, it's a matter of pushing this in and bring some last changes on the build process for the `px4_ros_com` package and should be good to go. In terms of general usage of RTPS, I think it's a matter of updating the the documentation for a more clear view of what's required and the current status, which is in my todo list as well.

bkueng commented on msg/templates/urtps/RtpsTopics.cpp.template:
```diff
@@ -88,7 +89,7 @@ void RtpsTopics::publish(uint8_t topic_ID, char data_buffer[], size_t len)
     {
 @[for topic in send_topics]@
-        case @(message_id(topic)): // @(topic)
+        case @(rtps_message_id(os.path.abspath('uorb_rtps_message_ids.json'), topic)): // @(topic)
```
This needs to parse a json file for every message. Can we instead pass an already parsed `dict` to the template?

TSC21 replied:
Makes sense. Will do that then.

bkueng commented on msg/tools/generate_microRTPS_bridge.py:
```diff
@@ -57,13 +57,21 @@ def get_absolute_path(arg_parse_dir):
 parser.add_argument("-t", "--topic-msg-dir", dest='msgdir', type=str, nargs=1, help="Topics message dir, by default msg/", default="msg")
+parser.add_argument("-b", "--uorb-templates-dir", dest='uorb_templates', type=str, nargs=1, help="uORB templates dir, by default msg_dir/templates/uorb_microcdr", default=default_uorb_templates_dir)
+parser.add_argument("-q", "--urtps-templates-dir", dest='urtps_templates', type=str, nargs=1, help="uRTPS templates dir, by default msg_dir/templates/urtps", default=default_urtps_templates_dir)
+parser.add_argument("-p", "--package", dest='package', type=str, nargs=1, help="Msg package naming, by default px4", default=default_package_name)
```
You can drop `nargs=1`, then you don't get a list.

TSC21 replied:
sure. thanks

bkueng commented on msg/tools/generate_microRTPS_bridge.py:
```diff
@@ -77,18 +85,26 @@ def get_absolute_path(arg_parse_dir):
 agent = args.agent
 client = args.client
+# If not specified, auto generate the idl filename_send_msgs
+if args.idl is None or args.idl == "":
```
If you add a `default=False` to `parser.add_argument("-i", "--no-idl"`, you can directly set `idl = args.idl`

TSC21 replied:
true. thanks

TSC21 commented:
> We could switch to yaml instead of json, as it is easier to edit and ROS uses it as well.

Sounds reasonable. Will use `yaml` instead.

dagar commented on msg/tools/uorb_rtps_message_ids.json:
```diff
@@ -0,0 +1,352 @@
+  "rtps": [
+    {
+      "msg": "actuator_armed",
+      "id": 1
```
Could we do this as a metadata field in the message itself? Alternatively something in the CMakeLists.txt might also make sense. https://github.com/PX4/Firmware/blob/master/msg/CMakeLists.txt#L35-L140
I'm worried about maintaining the list over time.

TSC21 commented:
@dagar seems we are missing `pyyaml` on OSX: http://ci.px4.io:8080/blue/organizations/jenkins/PX4_misc%2FFirmware-compile/detail/PR-10388/4/pipeline/42

TSC21 commented:
@bkueng everything addressed and now compiling. Thanks for the tips. Now we just need https://github.com/PX4/containers/pull/133 in.

TSC21 commented on msg/templates/urtps/RtpsTopics.h.template:
```diff
@@ -9,15 +9,24 @@
 topic_names = [single_spec.short_name for single_spec in spec]
 send_topics = [s.short_name for idx, s in enumerate(spec) if scope[idx] == MsgScope.SEND]
 recv_topics = [s.short_name for idx, s in enumerate(spec) if scope[idx] == MsgScope.RECEIVE]
+yaml_file = os.path.abspath('uorb_rtps_message_ids.yaml')
```
Makes sense. I will investigate further on how to do it.

TSC21 commented on src/modules/micrortps_bridge/micrortps_client/CMakeLists.txt:
```diff
@@ -71,14 +71,23 @@ if (NOT "${config_rtps_send_topics}" STREQUAL "" OR NOT "${config_rtps_receive_t
 		)
 
+	# Copy yaml file to the bin dir to be used by empy on the file generation
+	configure_file("${PX4_SOURCE_DIR}/msg/tools/uorb_rtps_message_ids.yaml"
```
Unless we set it manually on the template file or pass it through `globals` as you suggested, this was the only way I found, as the generator code makes a copy of the template and generates the code on the bin dir, so it's as the python script is run on the same dir, which leads to `os.path.abspath('uorb_rtps_message_ids.yaml')` to look on that same dir. I will look further for a solution here, but wouldn't want to spend much more time on it.

TSC21 commented on msg/tools/px_generate_uorb_topic_helper.py:
```diff
@@ -4,6 +4,9 @@
 # precompiled and thus message generation will be much faster
 
+import os
+import yaml
```
Ok sure.

TSC21 replied:
done

TSC21 commented:
> Looks good, I added a some remarks.

@bkueng I will try to address your remarks but this will also cause some changes on the `px4_ros_com` package I am developing, which would delay it's progress. Can these changes like the `'uorb_rtps_message_ids.yaml'` dir be addressed on another PR, or do you see them as required at this step?

TSC21 commented on msg/tools/generate_microRTPS_bridge.py:
```diff
@@ -77,20 +87,30 @@ def get_absolute_path(arg_parse_dir):
+    package = args.package
 else:
-    msg_files_receive = []
+    package = px_generate_uorb_topic_files.PACKAGE
```
done

TSC21 commented on msg/tools/generate_microRTPS_bridge.py:
```diff
@@ -77,20 +87,30 @@ def get_absolute_path(arg_parse_dir):
+if idl_dir != '':
+    idl_dir = get_absolute_path(args.idl_dir)
+else:
+    idl_dir = agent_out_dir + "/idl"
```
done

TSC21 commented:
@bkueng I addressed your comments except the ones related to `uorb_rtps_message_ids.yaml` which I think we can do in a new PR. I also did some beautification of the code according to PEP8. Is that ok with you?

TSC21 commented:
@bkueng I add two more options to the argparser. Can you approve and merge?

LorenzMeier commented:
Great! Can you please also send a documentation PR for folks who want to use this?

TSC21 commented:
> Great! Can you please also send a documentation PR for folks who want to use this?

Yes. But in this case, what would be the application context? Someone that wants to have a particular implementation of the agent on their system, with some specific agent templates and IDL's? (not including ROS2)

TSC21 commented:
@LorenzMeier had to rebase so to check if CI passes with the new docker container tags on master.

TSC21 commented:
@LorenzMeier the reason: https://github.com/PX4/Firmware/pull/10463#issuecomment-420555650

dagar commented:
Getting closer. Snapdragon needed an update as well. https://github.com/PX4/Firmware/pull/10468

TSC21 commented:
@dagar OSX is still giving us trouble...

dagar commented:
That's weird, on that machine (dronecode-macmini) I used this script to update the dependencies yesterday. https://github.com/PX4/Firmware/blob/master/Tools/setup/OSX.sh

Do you have a mac?

TSC21 commented:
> Do you have a mac?

I don't. I think this can be some clash between python stuff installed by pip and by homebrew.

TSC21 commented:
Anyone with a Mac that can have a test to https://github.com/PX4/Firmware/blob/master/Tools/setup/OSX.sh?

TSC21 commented:
@dagar can you please try the changes I put on my latest commit on `dronecode-macmini`? I am not sure about the exports, but they shouldn't be harmful. Also, can you test the output of `which -a pip` and `pip -V`? It may be that it's installing the packages in `/usr/local/lib/python2.7/site-packages`, so that path needs to be added to `PATH`.

TSC21 commented:
> That's weird, on that machine (dronecode-macmini) I used this script to update the dependencies yesterday. https://github.com/PX4/Firmware/blob/master/Tools/setup/OSX.sh

Can you check if pyyaml was correctly installed?

LorenzMeier commented:
At minimum this is required - but I'm still getting errors, as I had some of the packages already on my system - can fiddle with it later today.

```
diff --git a/Tools/setup/OSX.sh b/Tools/setup/OSX.sh
index d83723cd50..46611673ce 100755
--- a/Tools/setup/OSX.sh
+++ b/Tools/setup/OSX.sh
@@ -8,7 +8,7 @@ brew install px4-dev
 
 # python dependencies
 sudo easy_install pip
-pip install -r ${DIR}/requirements.txt
+sudo pip install -r ${DIR}/requirements.txt
 
 # Optional, but recommended additional simulation tools:
 brew install px4-sim
```

TSC21 commented:
@LorenzMeier any luck in the mean time?

TSC21 commented:
@MaEtUgR are you able to update the windows build in appveyor so CI can pass? (https://ci.appveyor.com/project/MaEtUgR/firmware/build/1458)
Thanks!

MaEtUgR commented:
Script update (tested to work with this pr): https://github.com/PX4/windows-toolchain/commit/6f752940919211953008cb2b08ba1243fb176a0c
I'm generating the new msi with multiple fixes at the moment. I need to continue scripting and make that fully automated on a cloud server.

TSC21 commented:
> I'm generating the new msi with multiple fixes at the moment. I need to continue scripting and make that fully automated on a cloud server.

Awesome! Thank you! Let me know when I can rebase this so to use the new msi file.

MaEtUgR commented:
@TSC21 Done: #10516 just cherry-pick my commit to change the toolchain version.

TSC21 commented:
Awesome! Thanks! We are just missing OSX now.

MaEtUgR commented:
@TSC21 I made #10516 independent such that you don't have to worry and can just rebase. Also it helps if every CI call uses the same file because it gets chached on appveyor.

thomasgubler commented:
@dagar this is blocking development of the rtsp messaging interface. Would it be possible to give @TSC21  ssh access to update the script?

TSC21 commented:
`dronecode-macmini` CI build fixed. Now missing `gus-macpro`. Already pinged @dogmaphobic about it.
I will then issue a PR with a small update on the `OSX.sh` setup script

TSC21 commented:
@bkueng FYI

TSC21 commented:
@dogmaphobic thanks for updating the machine! CI now passes. This is good to merge.

TSC21 commented:
@bkueng sure thing. Thanks for merging
